src/iso/cloud_init_iso.cpp

```diff
@@ -477,3 +562,83 @@ void mp::CloudInitIso::write_to(const Path& path)
     }
     pad_to_end(f);
 }
+
+void mp::CloudInitIso::read_from(const std::filesystem::path& fs_path)
```

Comment:
I think there are upsides of the `istream` approach, there is an element of decoupling and dependency injection. Because of that, the input can be "mocked"  by `istringstream` and furthermore provide a simpler way to test. However, there are also downsides, it is definitely a less simple and expressive function interface to the client code, and the client code also needs to construct the `istream` derive class object instead of letting the function handles internally. And more importantly, the open mode of the file can be anything instead of just `std::ios_base::in | std::ios::binary` which might cause some unexpected behavior. 

To conclude, I would say file system only use case should go with `std::filesystem::path`, and various source stream case (file stream, string stream network stream) should choose `istream`.

Reply:
`unique_ptr` part is seemingly a minor thing. How much `stringstream` can simplify things in **this particular case** is not clear to me, which is the major thing. 

Using `stringstream` to mock `ifstream` is a common practice and it has its advantages like easy to manipulate the **text stream** to create edge cases, being I/O free, etc in general cases. However, in the case of ISO format binary file, manually creating `stringstream` can be more challenging because of the binary file format and size of ISO file. Eventually, we may need to resort to writing a utility function to generate the `stringstream` we want. That will be essentially what `CloudInitIso::write_to` function is doing but with the `stringstream` instead of a `QFile`. 

The current unit tests does mock `read` calls. It also can default back the orginal implementation of the code. So it eventually can make one unit test framework something like `read` does the default things for N times and does the erroneous thing at N + 1th time, which are only two `EXPECT_CALL` calls for `read`, and as a result , it does not make the code of describing that scenario very long. 
 
The only gain I can see from this mocking in this particular case is the disk I/O free which lead to better run speed of the unit tests. But I do not think it worth the hassle.